### ISSUE TITLE
Limit resource usage of state_call

### ISSUE THREAD

**Issue by xlc** [paritytech/polkadot-sdk#1628]
Previous issue: https://github.com/paritytech/substrate/issues/12698

Do we have any resource limits of making runtime API calls via `state_call` RPC?

For Ethereum nodes, we can submit RPC call to ask it to arbitrary EVM computation but up to some hard coded gas limit, which could be like 10x of the block gas limit. This ensures a single RPC call cannot consume too much CPU on the node.

I think we need a similar mechanism for `state_call`. Make it accept an optional Weight which capped to like 10x block weight limit and interrupt the wasm function call when it is taking too long.

**Comment by pepyakin** [comment]
Related: https://github.com/paritytech/substrate/issues/10648

**Comment by bkchr** [comment]
Yeah, I think we should use the fuel mechanism and not `weight`. `weight` would require that we also start benchmarking the runtime apis.

**Comment by ggwpez** [comment]
>Make it accept an optional Weight which capped to like 10x block weight limit and interrupt the wasm function call when it is taking too long.

We can just use a tokio timeout? It is not deterministic, but should already prevent DoS vector.

**Comment by xlc** [comment]
That will be a simple starting point

**Comment by pepyakin** [comment]
I am not sure how that would work. Imagine that a thread enters in wasm and that wasm just loops indefinitely. When the timeout triggers it drops the future.

If you execute wasm, it is a blocking operation. That means that even if you armed a timeout, it won't ever trigger. It will elapse, but the future cannot be polled while it is blocked in the wasm code!

You can use [epochs](https://docs.rs/wasmtime/latest/wasmtime/struct.Engine.html#method.increment_epoch) but that would be pretty involved. [fuel](https://docs.rs/wasmtime/latest/wasmtime/struct.Config.html#method.consume_fuel) seems the most simple option here.
